datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapWriter.java

```diff
@@ -0,0 +1,216 @@
+    }
+  }
+
+  @Override public void onPageAdded(int blockletId, int pageId, ColumnPage[] pages)
+      throws IOException {
+    col2Ordianl.clear();
+    col2DataType.clear();
+    for (int colId = 0; colId < pages.length; colId++) {
+      String columnName = pages[colId].getColumnSpec().getFieldName().toLowerCase();
+      if (indexedColumns.contains(columnName)) {
+        col2Ordianl.put(columnName, colId);
+        DataType columnType = pages[colId].getColumnSpec().getSchemaDataType();
+        col2DataType.put(columnName, columnType);
+      }
+    }
+
+    // for each row
+    for (int rowId = 0; rowId < pages[0].getPageSize(); rowId++) {
+      // for each indexed column
+      for (int indexColId = 0; indexColId < indexedColumns.size(); indexColId++) {
+        String indexedCol = indexedColumns.get(indexColId);
+        byte[] indexValue;
+        if (DataTypes.STRING == col2DataType.get(indexedCol)
+            || DataTypes.BYTE_ARRAY == col2DataType.get(indexedCol)) {
+          byte[] originValue = (byte[]) pages[col2Ordianl.get(indexedCol)].getData(rowId);
+          indexValue = new byte[originValue.length - 2];
+          System.arraycopy(originValue, 2, indexValue, 0, originValue.length - 2);
+        } else {
+          Object originValue = pages[col2Ordianl.get(indexedCol)].getData(rowId);
+          indexValue = CarbonUtil.getValueAsBytes(col2DataType.get(indexedCol), originValue);
+        }
+
+        indexBloomFilters.get(indexColId).put(indexValue);
+      }
+    }
+  }
+
+  private void initDataMapFile() throws IOException {
+    String dataMapDir = genDataMapStorePath(this.writeDirectoryPath, this.dataMapName);
+    for (int indexColId = 0; indexColId < indexedColumns.size(); indexColId++) {
+      String dmFile = dataMapDir + File.separator + this.currentBlockId
+          + '.' + indexedColumns.get(indexColId) + BloomCoarseGrainDataMap.BLOOM_INDEX_SUFFIX;
+      DataOutputStream dataOutStream = null;
+      ObjectOutputStream objectOutStream = null;
+      try {
+        FileFactory.createNewFile(dmFile, FileFactory.getFileType(dmFile));
+        dataOutStream = FileFactory.getDataOutputStream(dmFile,
+            FileFactory.getFileType(dmFile));
+        objectOutStream = new ObjectOutputStream(dataOutStream);
+      } catch (IOException e) {
+        CarbonUtil.closeStreams(objectOutStream, dataOutStream);
+        throw new IOException(e);
+      }
+
+      this.currentDMFiles.add(dmFile);
+      this.currentDataOutStreams.add(dataOutStream);
+      this.currentObjectOutStreams.add(objectOutStream);
+    }
+  }
+
+  private void writeBloomDataMapFile() throws IOException {
+    for (int indexColId = 0; indexColId < indexedColumns.size(); indexColId++) {
+      BloomDMModel model = new BloomDMModel(this.currentBlockId, this.currentBlockletId,
+          indexBloomFilters.get(indexColId));
+      // only in higher version of guava-bloom-filter, it provides readFrom/writeTo interface.
```

Comment:
why we can not use higher version of guava? which component prevents us from introduce higher version? which version?

Reply:
![guava_bloom_dependency](https://user-images.githubusercontent.com/10445758/39085586-77c45104-45b7-11e8-8852-68c39a4734e6.PNG)


As shown above, hadoop-common use 11.0.2 version of guava. I've checked the interface of readFrom/writeTo of the bloomfilter provided by guava, it needs 24.* version. See https://github.com/google/guava/commit/62d17005a48e9b1044f1ed2d5de8905426d75299#diff-223d254389aa08fae7876742f8f07e8c for detailed information.

Since hadoop-common rely on this package, So I not going to replace it.

Reply:
ok